Source/WTF/wtf/AliasedRefPtr.h

```diff
@@ -0,0 +1,108 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+namespace WTF {
+
+template <typename VisibleType>
+class AliasedRefPtr {
+public:
+    explicit AliasedRefPtr(const AliasedRefPtr* other)
+        : m_ptr(other->m_ptr)
+        , m_ptrRef(other->m_ptrRef)
+        , m_ptrDeref(other->m_ptrDeref)
+    {
+        ASSERT(other);
+        ref();
+    }
+
+    template <typename RefCountedType>
+    explicit AliasedRefPtr(RefCountedType* ptr)
+        : m_ptr(ptr)
+        , m_ptrRef([](VisibleType* d) {
+            static_cast<RefCountedType*>(d)->ref();
+        })
+        , m_ptrDeref([](VisibleType* d) {
+            static_cast<RefCountedType*>(d)->deref();
+        })
+    {
+        ref();
+    }
+
+    explicit AliasedRefPtr(AliasedRefPtr&& other)
+        : m_ptr(std::exchange(other.m_ptr, nullptr))
+        , m_ptrRef(other.m_ptrRef)
+        , m_ptrDeref(other.m_ptrDeref)
+    {
+    }
+
+    explicit AliasedRefPtr(const AliasedRefPtr& other)
+        : m_ptr(other.m_ptr)
+        , m_ptrRef(other.m_ptrRef)
+        , m_ptrDeref(other.m_ptrDeref)
+    {
+        ref();
+    }
+
+    virtual ~AliasedRefPtr()
+    {
+        deref();
+    }
+
+    AliasedRefPtr& operator=(AliasedRefPtr&& other)
+    {
+        if (&other == this)
+            return *this;
+        m_ptr = std::exchange(other.m_ptr, nullptr);
+        m_ptrDeref = other.m_ptrDeref;
+        return *this;
+    }
+
+    explicit operator bool() const { return !!m_ptr; }
+    VisibleType* operator->() const { return m_ptr; }
+    VisibleType& operator*() const { ASSERT(m_ptr); return *m_ptr; }
+    operator VisibleType& () const { ASSERT(m_ptr); return *m_ptr; }
```

Comment:
It's incorrect to have a ref operator on a RefPtr class. I realize this is just a move out of a class, but this should be fixed before exposing it to clients.

Reply:
I've removed the conversion operator

Reply:
I've removed the entire class, see below